src/modules/userCreationRequestMethods.ts

```diff
@@ -372,6 +372,52 @@ export const getUserCreationRequestQuery = async (id: string): Promise<UserCreat
     };
 };
 
+export const findUsersByEmails = async (
+    emails: string[],
+): Promise<Map<string, ExtractTypeFromGenerated<KyselyUser>>> => {
+    if (!emails.length) return new Map();
+
+    const users = await db
+        .with('user_emails', (qb) =>
+            qb
+                .selectFrom('User')
+                .select(['User.id as userId', 'User.email as email'])
+                .where('User.email', 'in', emails)
+                .unionAll(
+                    qb
+                        .selectFrom('UserServices')
+                        .innerJoin('User', 'User.id', 'UserServices.userId')
+                        .select(['UserServices.userId as userId', 'UserServices.serviceId as email'])
+                        .where('UserServices.serviceId', 'in', emails)
+                        .where('UserServices.serviceName', 'in', [
+                            ExternalServiceName.WorkEmail,
+                            ExternalServiceName.PersonalEmail,
+                            ExternalServiceName.Email,
+                        ]),
+                ),
+        )
+        .selectFrom('User')
+        .innerJoin('user_emails', 'user_emails.userId', 'User.id')
+        .selectAll('User')
+        .select([sql<string[]>`array_agg(distinct user_emails.email)`.as('emails')])
```

Comment:
You can write this query without subselect, you can use one `leftJoin` directive for join `UserService` table to `User` table